src/index.js

```diff
@@ -95,48 +111,195 @@ class Masq {
+    })
+
+    sw.on('close', () => {
+      hub.close()
+    })
+
+    sw.on('disconnect', (peer, id) => {
+      sw.close()
+      hub.close()
+    })
+
+    const _handleData = async (data, peer) => {
+      const json = JSON.parse(data)
+
+      switch (json.msg) {
+        case 'sendProfilesKey':
+          // check challenges
+          if (json.challenge !== this.challenge) {
+            // This peer may be malicious, close the connection
+            sw.close()
+            hub.close()
+          } else {
+            // db creation
+            const db = hyperdb(rai('masq-profiles'), Buffer.from(json.key, 'hex'), { valueEncoding: 'json' })
+            await dbReady(db)
+
+            // Store
+            this.dbs.profiles = db
+
+            peer.send(JSON.stringify({
+              msg: 'replicationProfilesStarted'
+            }))
+            // db replication
+            this._startReplication(db)
+          }
+          break
+        default:
+          break
+      }
+    }
+  }
+
+  /**
+   * After the masq-profiles replication, the right profil is chosen,
+   * the next steps are :
+   * - getting the hyperdb key from masq
+   * - request write authorization by sending the local key
+   */
+  exchangeDataHyperdbKeys () {
+    // Subscribe to channel for a limited time to sync with masq
+    if (!this.profile) throw (new Error('No profile selected'))
+    const hub = signalhub(this.channel, [HUB_URL])
+    let sw = null
 
     if (swarm.WEBRTC_SUPPORT) {
-      this.sw = swarm(this.hub)
+      sw = swarm(hub)
     } else {
-      this.sw = swarm(this.hub, { wrtc: require('wrtc') })
+      sw = swarm(hub, { wrtc: require('wrtc') })
     }
 
-    this.sw.on('peer', (peer, id) => {
+    sw.on('peer', (peer, id) => {
       // check challenges
-      peer.on('data', data => this._handleData(data))
+      peer.on('data', data => _handleData(data, peer))
+    })
+
+    sw.on('close', () => {
+      hub.close()
+    })
+
+    sw.on('disconnect', (peer, id) => {
+      sw.close()
+      hub.close()
+    })
+
+    const _handleData = async (data, peer) => {
+      const json = JSON.parse(data)
+
+      switch (json.msg) {
+        case 'sendDataKey':
+          if (json.challenge !== this.challenge) {
+            // This peer may be malicious, close the connection
+            sw.close()
+            hub.close()
+          } else {
+            // db creation and replication
+            const db = hyperdb(rai(this.profile), Buffer.from(json.key, 'hex'), { valueEncoding: 'json' })
+            await dbReady(db)
+            // Store
+            this.dbs[this.profile] = db
+
+            peer.send(JSON.stringify({
+              msg: 'requestWriteAccess',
+              key: db.local.key.toString('hex')
+            }))
+          }
+          break
+        case 'ready':
+          // Masq must send ready after the authorization
+          this._startDataReplication()
+          sw.close()
```

Comment:
missing `hub.close()` too

Reply:
not useful because hub.close will be called automatically